### PR TITLE
CLI summaries for actions without annotations

### PR DESCRIPTION
**Pull request by underwoodb-sd-ibm**
This update is intended to clarify the printed summary when one uses `wsk [ENTITY] [ENTITY_NAME] get --summary`.

Here is an example of what the output of this update should look like:

For entities with multiple parameters and no description:
```
$ wsk action get hola -s
action /guest/hola: Returns a result based on parameters testA, testB and testC
   (parameters: testA, testB, testC)
```

For entities with a single parameter and no description:
```
$ wsk action get hola -s
action /guest/hola: Returns a result based on parameter testA
   (parameters: testA)
```

And entities with no parameters and no description:
```
$ wsk action get hello -s
action /guest/hello
   (Warning: /guest/hello has no annotations)
```

 Closes https://github.com/apache/incubator-openwhisk/issues/2270

### CODE REVIEW

**Comment by underwoodb-sd-ibm** [comment]
@mdeuser, @dubee - I still need to update docs, but this is my rough draft for the PR. Still have some other functionality I'd like to add in that @rabbah talks about in the original issue. Just wanted to make you aware of this PR. Thanks!

**Comment by rabbah** [comment]
most actions will not have annotations - i suggest dropping the "warning" prefix.

**Comment by rabbah** [comment]
when no annotation is present, you can still construct a summary of the bound parameters - which should be part of the summary anyway imo.

File: tools/cli/go-whisk-cli/commands/util.go [hunk 1]
```diff
@@ -427,17 +427,31 @@ func printRuleSummary(rule *whisk.Rule) {
...
 func printEntitySummary(entityType string, fullName string, description string, params string) {
-    if len(description) > 0 {
+    if len(description) > 0 && len(params) > 0 {
         fmt.Fprintf(color.Output, "%s %s: %s\n", boldString(entityType), fullName, description)
+    } else if len(params) > 0 {
+        descriptionFromParams := buildParamString(params)
+        fmt.Fprintf(color.Output, "%s %s: %s\n", boldString(entityType), fullName, descriptionFromParams)
     } else {
         fmt.Fprintf(color.Output, "%s %s\n", boldString(entityType), fullName)
+        fmt.Fprintf(color.Output, "   (%s: %s has no annotations)\n", boldString("Warning"), fullName)
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/util.go]: The text added here must be externalized. Also, maybe "does not have any annotations".

Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/util.go]: "does not have annotations."

Comment (underwoodb-sd-ibm) [inline tools/cli/go-whisk-cli/commands/util.go]: When you say "externalized", I'm assuming you mean put the text into a variable and just call the variable in the print statement?

**Comment by underwoodb-sd-ibm** [comment]
PG4 585

**Comment by underwoodb-sd-ibm** [comment]
PG2 1852 - Passed

File: tests/src/test/scala/system/basic/WskBasicTests.scala [hunk 2]
```diff
@@ -130,6 +130,84 @@ class WskBasicTests
...
+                pkgParamBound -> JsString("Bound"),
+                pkgParamOverlap -> JsString("Bound And Annotated"))
+
+            assetHelper.withCleaner(wsk.pkg, pkgBoundParams) {
+                (pkg, _) =>
+                    pkg.create(pkgBoundParams, annotations = pkgAnnots, parameters = pkgParamsBound)
+            }
+
+            val pkgStdoutBound = wsk.pkg.get(pkgBoundParams, summary = true).stdout
+            val ns_regex_list = wsk.namespace.list().stdout.trim.replace('\n', '|')
```
Comment (dubeejw) [inline tests/src/test/scala/system/basic/WskBasicTests.scala]: ns_regex_list is not used in this test.

File: tests/src/test/scala/system/basic/WskBasicTests.scala [hunk 3]
```diff
@@ -520,6 +690,84 @@ class WskBasicTests
...
+                trgParamBound -> JsString("Bound"),
+                trgParamOverlap -> JsString("Bound And Annotated"))
+
+            assetHelper.withCleaner(wsk.trigger, trgBoundParams) {
+                (trigger, _) =>
+                    trigger.create(trgBoundParams, annotations = trgAnnots, parameters = trgParamsBound)
+            }
+
+            val trgStdoutBound = wsk.trigger.get(trgBoundParams, summary = true).stdout
+            val ns_regex_list = wsk.namespace.list().stdout.trim.replace('\n', '|')
```
Comment (dubeejw) [inline tests/src/test/scala/system/basic/WskBasicTests.scala]: ns_regex_list not used.

**Comment by dubeejw** [comment]
@underwoodb-sd-ibm, looks like Travis failed here.

File: tests/src/test/scala/system/basic/WskBasicTests.scala [hunk 4]
```diff
@@ -410,6 +487,98 @@ class WskBasicTests
...
+                    action.create(nameBoundParams, defaultAction, annotations = annots, parameters = paramsBound)
+            }
+            assetHelper.withCleaner(wsk.action, nameFinalParams) {
+                (action, _) =>
+                    action.create(nameFinalParams, defaultAction, annotations = annotsFinal, parameters = paramsBound)
+            }
+
+            val stdoutBound = wsk.action.get(nameBoundParams, summary = true).stdout
+            val stdoutFinal = wsk.action.get(nameFinalParams, summary = true).stdout
+            val ns_regex_list = wsk.namespace.list().stdout.trim.replace('\n', '|')
```
Comment (dubeejw) [inline tests/src/test/scala/system/basic/WskBasicTests.scala]: `ns_regex_list` is not used.

File: docs/packages.md [hunk 5]
```diff
@@ -42,15 +42,21 @@ Several packages are registered with OpenWhisk. You can get a list of packages i
...
+     (parameters: dbname, id, params)
+   action /whisk.system/cloudant/write: Write document in database
+     (parameters: dbname, doc)
    feed   /whisk.system/cloudant/changes: Database change feed
+     (parameters: dbname, filter, query_params)
+  ...
   ```
 
-  This output shows that the Cloudant package provides two actions, `read` and `write`, and one trigger feed called `changes`. The `changes` feed causes triggers to be fired when documents are added to the specified Cloudant database.
+  **Note**: Parameters listed under the package with a prefix `*` are predefined, bound parameters. Parameters without a `*` are those listed under the [annotations](./annotations.md) for each entity. Any entity listed under a package inherits specific bound parameters from the package. To view the list of known parameters of an entity belonging to a package, you will need to run a `get --summary` of the individual entity.
```
Comment (dubeejw) [inline docs/packages.md]: Need to explain what `**` denotes in the docs.

**Comment by underwoodb-sd-ibm** [comment]
PG3 980

**Comment by underwoodb-sd-ibm** [comment]
PG3 980 - Passed

**Comment by dubeejw** [comment]
@csantanapr, want to merge?

File: tests/src/test/scala/system/basic/WskBasicTests.scala [hunk 6]
```diff
@@ -130,6 +130,83 @@ class WskBasicTests
...
     }
 
+    it should "create, and get a package summary without a description and/or parameters" in withAssetCleaner(wskprops) {
+        (wp, assetHelper) =>
+            val pkgNoDesc = "pkgNoDesc"
+            val pkgNoParams = "pkgNoParams"
+            val pkgNoDescOrParams = "pkgNoDescOrParams"
+            val pkgDesc = "Package description"
+            val descFromParams = "Returns a result based on parameters"
+            val ns_regex_list = wsk.namespace.list().stdout.trim.replace('\n', '|')
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicTests.scala]: this should be `wsk.namespace.whois()`.

File: docs/packages.md [hunk 7]
```diff
@@ -59,10 +65,12 @@ Several packages are registered with OpenWhisk. You can get a list of packages i
   ```
   ```
   action /whisk.system/cloudant/read: Read document from database
-     (params: dbname includeDoc id)
+     (parameters: *apihost, *bluemixServiceName, *dbname, *host, *id, params, *password, *username)
```
Comment (rabbah) [inline docs/packages.md]: i'd remove `bluemixServiceName` here.

File: tests/src/test/scala/system/basic/WskBasicTests.scala (hunk 6 above)
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicTests.scala]: i'd rename `ns_regex_list` to just namespace.

File: tests/src/test/scala/system/basic/WskBasicTests.scala [hunk 8]
```diff
@@ -410,6 +487,97 @@ class WskBasicTests
...
+            }
+            assetHelper.withCleaner(wsk.action, actNameNoDescOrParams) {
+                (action, _) =>
+                    action.create(actNameNoDescOrParams, defaultAction)
+            }
+
+            val stdoutNoDesc = wsk.action.get(actNameNoDesc, summary = true).stdout
+            val stdoutNoParams = wsk.action.get(actNameNoParams, summary = true).stdout
+            val stdoutNoDescOrParams = wsk.action.get(actNameNoDescOrParams, summary = true).stdout
+            val ns_regex_list = wsk.namespace.list().stdout.trim.replace('\n', '|')
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicTests.scala]: same as above.
there's others, i trust you'll address all instances.

**Comment by underwoodb-sd-ibm** [comment]
PG4 653

**Comment by underwoodb-sd-ibm** [comment]
PG2 1930 - Passed

**Comment by underwoodb-sd-ibm** [comment]
@rabbah - could you review the changes I made and see if they are satisfactory for your review? Thanks!

File: tests/src/test/scala/system/basic/WskBasicTests.scala [hunk 9]
```diff
@@ -130,6 +130,83 @@ class WskBasicTests
             stdout should include regex (s"(?i)action /$ns/$packageName/$actionName: Action description\\s*\\(parameters: paramName1, paramName2\\)")
     }
 
+    it should "create, and get a package summary without a description and/or parameters" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicTests.scala]: I move move these tests out of the `system` package to wsk basic usage tests.

**Comment by rabbah** [comment]
looks good - just please move the tests out of the system package. generally usage tests don't go in the system package.

**Comment by rabbah** [comment]
don't need to re-pg.

**Comment by underwoodb-sd-ibm** [comment]
Thanks, @rabbah! Moving tests for all the PR's now 👍

**Comment by underwoodb-sd-ibm** [comment]
@rabbah, test has been moved to WskBasicUsageTests. Please confirm where I put it in the test is satisfactory, and if so, I'll tag Carlos for merging.

**Comment by dubeejw** [comment]
@underwoodb-sd-ibm, looks like there are still tests in WskBasicTests.

**Comment by underwoodb-sd-ibm** [comment]
@dubeejw - fixed :)

**Pull request merged**